File: frontend/src/api/operations.ts
import { UseQueryResult } from "react-query";
import { User } from "./dataTypes";
import useUser from "./user";

const Operations = {
	page_events: "page_events",
	page_staff: "page_staff",
	page_debug: "page_debug",
	page_users: "page_users",
	page_fonts: "page_fonts",
	page_managers: "page_managers",

	get_staff_event: "get_staff_event",

	add_event: "add_event",
	add_ticket: "add_ticket",
	add_staff: "add_staff",
	add_font: "add_font",
	add_manager: "add_manager",

	change_ticket_types: "change_ticket_types",
	change_ticket: "change_ticket",
	change_event: "change_event",
	change_staff_event: "change_staff_event",
	change_staff: "change_staff",

	delete_event: "delete_event",
	delete_ticket: "delete_ticket",
	delete_staff: "delete_staff",
	delete_manager: "delete_manager",
}

export type Operation = keyof typeof Operations;

export default function hasPermission(user: UseQueryResult<User, unknown>, operation: Operation)
{
	return !!(user.data?.auth && user.data?.operations.includes(operation));
}

export function useHasPermission(operation: Operation)
{
	const user = useUser();
	return !!(user.data?.auth && user.data?.operations.includes(operation));
}

File: backend/data/_operations.py
from bafser import OperationsBase


class Operations(OperationsBase):
    # pages
    page_events = ("page_events", "Страница мероприятия")
    page_staff = ("page_staff", "Страница сотрудников")
    page_debug = ("page_debug", "Страница отладки")
    page_debug_events = ("page_debug_events", "Страница отладки событий")
    page_debug_users = ("page_debug_users", "Страница отладки пользователей")
    page_fonts = ("page_fonts", "Страница шрифтов")
    page_managers = ("page_managers", "Страница организаторов")

    # get
    get_staff_event = ("get_staff_event", "Просмотр сотрудников на мероприятии")

    # add
    add_event = ("add_event", "Создание мероприятий")
    add_ticket = ("add_ticket", "Добавление билетов")
    add_staff = ("add_staff", "Добавление сотрудников")
    add_any_image = ("add_any_image", "Добавление любых картинок")
    add_font = ("add_font", "Добавление шрифтов")
    add_manager = ("add_manager", "Добавление организаторов")

    # change
    change_ticket_types = ("change_ticket_types", "Изменение типов билетов")
    change_ticket = ("change_ticket", "Изменение билетов")
    change_event = ("change_event", "Изменение мероприятий")
    change_staff = ("change_staff", "Изменение сотрудников")
    change_staff_event = ("change_staff_event", "Изменение сотрудников на мероприятии")

    # delete
    delete_event = ("delete_event", "Удаление мероприятий")
    delete_ticket = ("delete_ticket", "Удаление билетов")
    delete_staff = ("delete_staff", "Удаление сотрудников")
    delete_manager = ("delete_manager", "Удаление организаторов")
